File: minimumCapacity.py
 class Solution:
    def isValid(self, weights, capacity, days):
        total_sum = 0
        count = 1
        for weight in weights:
            if total_sum + weight > capacity:
                count += 1
                total_sum = 0

            total_sum += weight

        return count <= days

    def minimumCapacity(self, weights, days):
        n = len(weights)

        lo = 0
        hi = 0

        for weight in weights:
            lo = max(lo, weight)
            hi += weight

        while lo <= hi:
            mid = lo + (hi - lo) // 2

            if self.isValid(weights, mid, days):
                hi = mid
            else:
                lo = mid + 1

        return lo

File: minimumCapacity.cpp
#include<bits/stdc++.h>
using namespace std;
#define int long long 
#define fio ios_base::sync_with_stdio(false);

 class Solution{
    
     public:

         bool isValid(vector<int>&weights, int capacity, int days){
               int sum = 0;
               int cnt = 1;
               for(auto it: weights){
                   if(sum + it > capacity){
                      cnt++;
                      sum = 0;
                   } 

                   sum += it;
               }

               return cnt <= days;
         }
        
         int minimumCapacity(vector<int>&weights, int days){
              int n = weights.size();

              int lo = 0, hi = 0;

              for(auto it: weights){
                   lo = max(lo, it);
                   hi += it;
              }

              while(lo <= hi){
                  int mid = lo + (hi-lo)/2;

                  if(isValid(weights, mid, days)){
                       hi = mid;
                  }else{
                      lo = mid+1;
                  }
              }

              return lo;
         }
 };     


 
